### DISCUSSION TITLE
What is the purpose of `IWindowCreator`?

### DISCUSSION THREAD

**Discussion by Ghostbird** [dotnet/maui#26327] [Q&A] [answered]
I recently upgraded my MAUI 8 app to MAUI 9. One of the things I had to change was that I had to replace the `Application.MainPage = ` with an override of `Application.CreateWindow`.

So I overrode that function with the logical implementation:

```csharp
  protected override Window CreateWindow(IActivationState? activationState)
  {
    ArgumentNullException.ThrowIfNull(activationState);
    var window = base.CreateWindow(activationState);
    window.Page = activationState.Context.Services.GetRequiredService<AppShell>();
    return window;
  }
```

However, to facilitate this, I must also register an implementation of `IWindowCreator` with a trivial implementation.

```csharp
public class WindowCreator : IWindowCreator
{
	public Window CreateWindow(Application app, IActivationState? activationState) => new();
}
```

This works perfectly, but it leaves me with questions.

1. Why is no mention made that you must register an implementation of `IWindowCreator`?
2. Why is no default implementation of `IWindowCreator` provided?
3. Or if no default implementation of `IWindowCreator` is meant to be provided, why does the `CreateWindow` method not fall back to the trivial implementation: `return new()`?

The current state of the code seems incomplete, and ready for further development. I saw StackOverflow answers that override `CreateWindow` without calling the base class, or with an implementation that first creates a window and adds it to `Windows`. In doing so, you respectively; discard all the built-in guards, or create a window that will immediately be discarded once `IWindowCreator` is registered properly in DI. The current state of the MAUI code will lead naive programmers to write code like that, and suggest it on StackOverflow. I expect MAUI to later complete this incomplete mechanism, and that will cause lots of failures in such currently suggested implementations of `CreateWindow`.

**Comment by MartyIX** [comment]
I can see that https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/windows?view=net-maui-9.0#decouple-window-management-from-the-app-class mentions that one must register the creator:

```cs
builder.Services.AddSingleton<IWindowCreator, WindowCreator>();
```

Perhaps you can file an issue here https://github.com/dotnet/docs-maui/issues to ask for more details or create a PR to add information that is not obvious.

**Reply by Ghostbird** [reply]
Thanks for pointing that out. I get that it's for decoupling, and I'm very happy with that. Creating a docs issue might be a good idea. From a point of decoupling, I'm mostly unclear whether this is an `IWindowCreator` or an `IMainPageWindowCreator` the latter is what the documentation suggest.

In my current code I've used it to decouple the `Window` creation from the app using the `IWindowCreator`. However, given the name, I couldn't decouple the main page creation from the app. After all, `IWindowCreator` suggests that it will be used to create every window, not just the window that shows the main page.

**Reply by MartyIX** [reply]
> From a point of decoupling, I'm mostly unclear whether this is an `IWindowCreator` or an `IMainPageWindowCreator` the latter is what the documentation suggest.

I think that it's simply `IWindowCreator` and not `IMainPageWindowCreator`. The naming is IMHO as follows:

* For _window_ think of a Windows window.
* For _page_ think of some content in an actual operating system window.

`IWindowCreator` should be called for all windows you create but not for all pages you create.

> I recently upgraded my MAUI 8 app to MAUI 9. One of the things I had to change was that I had to replace the Application.MainPage = with an override of Application.CreateWindow.

But the deprecation warning recommends something like:

```cs
Application.Windows[0].MainPage = something
```

doesn't it? Is it not possible in your use case?

**Comment by Ghostbird** [comment, marked as answer]
I'm currently using this approach where I register the class below as a singleton `IWindowCreator`.
I improved upon some problems I recognised in the documented approach:

- Decouples main page lifetime from the application and its main window entirely. E.g. it can be called to recreate a main window without recreating the main page.
    Think of apps like Spotify and Discord that close their main window, but keep running, and leave some method to create a new main window again, that shows the page as it were when you closed it.
- Can be called to create subsequent non-main windows that will not show the main page.

```csharp
public class WindowCreator : IWindowCreator
{
  public Window CreateWindow(Application app, IActivationState? activationState)
  {
    var window = new Window();
    if (!app.Windows.Any())
    {
      // First window created shows the main page.
      // Note: Your main page class will most likely be different.
      window.Page = activationState?.Context.Services.GetRequiredService<AppShell>();
    }
    return window;
  }
}
```

If using this approach, do **not** set `Application.MainPage`, nor override `Application.CreateWindow`.
